### ISSUE TITLE
InternalError exception

### ISSUE THREAD

**Issue by Caulm** [bsdayo/BingChat#15]
I can often catch this exception.
`InternalError: Unhandled Dependency Failure: DeepLeo`

The text is part of the response message, which the websocket client received from the server. (see below)
```JSON
{
    "type": 2,
    "invocationId": "1",
    "item": {
        "firstNewMessageIndex": null,
        "defaultChatName": null,
        "conversationId": "51D|BingProdUnAuthenticatedUsers|E7D ... 284",
        "requestId": "88e6dd8c-2cae-49e4-aef4-223d27f69487",
        "telemetry": { },
        "result": {
            "value": "InternalError",
            "message": "Unhandled Dependency Failure: DeepLeo",
            "error": "Unhandled Dependency Failure: DeepLeo\n ---> ServiceClient failure for DeepLeo",
            "exception": "Microsoft.TuringBot.Common.ServiceInternalError: Unhandled Dependency Failure: DeepLeo\r\n ---> Microsoft.TuringBot.Common.ServiceClientException: ServiceClient failure for DeepLeo\r\n ---> System.Net.Http.HttpRequestException: Failed to call \"DeepLeo\" at \"SelectedByPapyrusLoadBalancer\"...(some call stack information)...",
            "serviceVersion": "20230518.47"
        }
    }
}
```

It seems to appear once in every 5 questions on average. If I asked the same question again, it can answer. I'm not sure if it's just for me, or if it's a network issue.

If you have any solutions or clues, please let me know. Thank you.

**Comment by Caulm** [comment]
I didn't notice the strange conversationId. it has always been like this to me.

Updating cookies does not work. But it becomes normal when I use Edge to ask questions, like: `"conversationId":"51D|BingProd|73E ... 3EA"`

If the issue is related to the Id, then something may have been wrong in `CreateConversation` process? I'll try some debugging.

**Comment by neon-sunset** [comment]
Well, the API is certainly more complicated than expected heh. While I working on response streaming, here are a few findings you may want to know (will be editing msg):
- BingChat uses `{\"type\":6}\u001e` msgs as what most likely is keep-alive and both server and client send these over `ChatHub` every 16 seconds
- Something makes websocket session server output malformed in the implementation of the library, and it is even recorded when browsing chat history, in the network it looks like it includes chunks of the client-sent messages. Related to keep-alive?

**Comment by neon-sunset** [comment]
SignalR implementation works (but not the "advanced" protocol negotiation) and does so well.
Basically:
- `update` is a client method that server pushes response updates to
- `chat` is a streaming server method invocation

SignalR update will be a part of response streaming PR.

After that, we'll need to try reproducing the issue once again. If persists, then my next bet is it has to do with not properly using some load-balancing mechanism and, if re-connection happens, landing at a different node (!pay attention to `SelectedByPapyrusLoadBalancer` in stacktrace!).

Notes: protocol negotiation starts and we even receive a specific connection ID but something happens during negotiation process which causes ChatHub to respond with 502 upon trying to establish connection with `wss://sydney.bing.com/sydney/ChatHub?id=...`.

**Comment by Caulm** [comment]
> Well, the API is certainly more complicated than expected heh. While I'm working on response streaming, here are a few findings you may be interested in (will be editing msg):
> 
> * BingChat uses `{\"type\":6}\u001e` msgs as what most likely is keep-alive and both server and client send these over `ChatHub` every 16 seconds. UPD: SignalR keepalive ping-pong logic? I should read about it more;
> * Something makes websocket session server output malformed in the implementation of the library before we get to message `{ type: 2 }`, and it is even recorded when browsing chat history in Edge, in the network it looks like it includes chunks of the client-sent messages. Related to keep-alive?
> 
> Todo:
> 
> * Look into straight up using SignalR client instead of hand-rolling websocket code (not trusting the library used initially 😅)
> 
> p.s.: at least this is JSON instead of MsgPack or Protobuf
> 
> UPD: The guess was correct. See https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/SignalR.Common/src/Protocol/HubProtocolConstants.cs#L14 It's definitely worth rewriting to SignalR which should fix any possible transport issues. I'm currently looking at BingChat's widget TS/JS code to deduce correct hub builder configuration. That way, we can match the shape of the official API decently well.

Thanks for suggestions.

I once encountered message `{type: 7}`, which caused the client getting stuck. Using SignalR is definitely a significant improvement for this and other potential transport issues.

The Edge browser can record our chat history, this maybe related to an `autosave` option? It seems to take effect after these options are added.
https://github.com/bsdayo/BingChat/blob/8f85b1c351dd39f5d99d07deba1bb2cf03bc1207/src/BingChat/BingChatConstants.cs#L22

**Comment by neon-sunset** [comment]
Type 7 indicates session termination: https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/SignalR.Common/src/Protocol/HubProtocolConstants.cs#L44

As for `autosave` - maybe? I have decided to take on issues one at a time: SinalR, then response streaming, then further reverse engineering of option sets and potential issues with sticky-sessions / load-balancing (if true) if network issues continue.

It is yet unknown the degree of control afforded by OptionSets, whether they are purely informational in nature, how back-end validates them, etc.

Consider the following commit: https://github.com/acheong08/EdgeGPT/commit/f536e79429449c62d0819024f12dc534d77780d6
Apparently, there is a way to select new vs old model (with new model often just sucking ass on prompts that used to work very reliably), but when I tried a few I did not notice significant gap you would see when comparing GPT-3.5 vs 4 side-by-side, which means they don't work in such a straightforward way.

All in all, I think we need to start documenting findings somewhere.
